Add orchestrator for supporting multiple processes

dineshg13 opened the pull request:
changes necessary to resolve https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/197 . 
This PR  moves the logic from `main`, and creates a new orchestrator pkg for handling multiple processes.

dineshg13 commented:
@edeNFed This PR is working on baremetal and as sidecar for a pod. But when i run same as a deployment in k8s, am getting error below error. Can you please help with same. 

```
{"level":"info","ts":1687914965.2464294,"caller":"cli/main.go:60","msg":"starting Go OpenTelemetry Agent ..."}
{"level":"info","ts":1687914965.2465029,"caller":"cli/main.go:62","msg":"Establishing connection to OTLP receiver ..."}
{"level":"info","ts":1687914967.2483935,"caller":"orchestrator/orchestrator.go:151","msg":"No go process not found yet, trying again in 2 seconds"}
{"level":"info","ts":1687914969.2487526,"caller":"orchestrator/orchestrator.go:151","msg":"No go process not found yet, trying again in 2 seconds"}
{"level":"info","ts":1687914971.24835,"caller":"orchestrator/orchestrator.go:151","msg":"No go process not found yet, trying again in 2 seconds"}
{"level":"info","ts":1687914973.247508,"caller":"orchestrator/orchestrator.go:151","msg":"No go process not found yet, trying again in 2 seconds"}
{"level":"info","ts":1687914975.2487097,"caller":"orchestrator/orchestrator.go:97","msg":"Add auto instrumetors","pid":2247,"serviceName":"sample-app"}
{"level":"info","ts":1687914975.250355,"caller":"ptrace/ptrace_linux.go:102","msg":"retry attaching thread","tid":2247,"retryCount":1,"limit":10,"error":"no such process"}
{"level":"info","ts":1687914975.2504175,"caller":"ptrace/ptrace_linux.go:102","msg":"retry attaching thread","tid":2248,"retryCount":1,"limit":10,"error":"no such process"}
{"level":"info","ts":1687914975.2504244,"caller":"ptrace/ptrace_linux.go:102","msg":"retry attaching thread","tid":2249,"retryCount":1,"limit":10,"error":"no such process"}
{"level":"info","ts":1687914975.2504296,"caller":"ptrace/ptrace_linux.go:102","msg":"retry attaching thread","tid":2250,"retryCount":1,"limit":10,"error":"no such process"}
{"level":"info","ts":1687914975.2504346,"caller":"ptrace/ptrace_linux.go:102","msg":"retry attaching thread","tid":2251,"retryCount":1,"limit":10,"error":"no such process"}
{"level":"info","ts":1687914975.25044,"caller":"ptrace/ptrace_linux.go:102","msg":"retry attaching thread","tid":2252,"retryCount":1,"limit":10,"error":"no such process"}
{"level":"info","ts":1687914975.250455,"caller":"ptrace/ptrace_linux.go:102","msg":"retry attaching thread","tid":2247,"retryCount":2,"limit":10,"error":"no such process"
```

damemi commented on cli/main.go:
```diff
@@ -15,18 +15,41 @@
 
+var (
+	// Controller-local reference to the auto-instrumentation release version.
+	releaseVersion = auto.Version()
```
`releaseVersion` is only used in the line below, do we need this variable?

damemi commented on pkg/process/args.go:
```diff
@@ -40,12 +46,14 @@ func (t *TargetArgs) Validate() error {
-	if exists {
+	if val, exists := os.LookupEnv(ExePathEnvVar); exists {
+		result := &TargetArgs{}
 		result.ExePath = val
+		serviceName, exists := os.LookupEnv(otelServiceNameEnvVar)
+		if exists {
+			result.ServiceName = serviceName
+		}
+		return result
 	}
-
-	return result
+	return nil
```
you return `nil` here if the exePath isn't set, which is invalid config, but then you can't call `Validate()` to catch it

maybe change this like:

```go
    result := &TargetArgs{}
	if val, exists := os.LookupEnv(ExePathEnvVar); exists {
		result.ExePath = val
	}
    if serviceName, exists := os.LookupEnv(otelServiceNameEnvVar); exists {
		result.ServiceName = serviceName
	}
	return result
```

breaking the `serviceName` check out of the nested exePath check is a little easier to read too. you could also fail earlier if exePath isn't set, but we have Validate() functions for that anyway

damemi replied:
what about an explicit field in `targetArgs` that says you want to monitor all services (eg `targetArgs.allTargets = true`) instead of relying on `nil` to imply that's what you're doing?

dineshg13 replied:
That is good idea. Will add it .

damemi commented on pkg/instrumentors/manager.go:
```diff
@@ -30,26 +30,24 @@ import (
 		instrumentors:  make(map[string]Instrumentor),
 		done:           make(chan bool, 1),
-		incomingEvents: make(chan *events.Event),
+		incomingEvents: make(chan *events.Event, 10),
```
why is this buffered to 10?

dineshg13 replied:
am surprised that we didn't had buffer before. Without buffer, the process would wait for GRPC request to complete.  This would impact performance of the process. I choose 10 at random, but ideally we should make this configurable.

damemi replied:
cool, yeah maybe some benchmarking tests would be good to have at some point

dineshg13 replied:
Added https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/334

damemi commented on pkg/opentelemetry/controller.go:
```diff
@@ -98,35 +77,23 @@ func (c *Controller) convertTime(t int64) time.Time {
-
-	ctx := context.Background()
+func NewController(
+	ctx context.Context,
+	serviceName string,
+	exporter sdktrace.SpanExporter,
```
not a blocker yet since this isn't stable but we might want to consider a different approach to passing params in if this list could grow

dineshg13 replied:
Updated . Added controller settings.

damemi commented on pkg/process/discover.go:
```diff
@@ -91,30 +93,55 @@ func (a *Analyzer) findProcessID(target *TargetArgs) (int, error) {
+			if _, ok := a.ignoreProcesses[exe]; ok {
+				continue
+			}
+			if target != nil && !strings.Contains(string(exeFullPath), target.ExePath) {
```
If I'm running in single-target mode, will this keep looping through processes after the target exe has been found?

dineshg13 replied:
Yes. Currently, once we have the pid and start monitoring the process. If the process dies, auto-instrumentation doesn't try to find the new pid. It doesn't allow us to run on bare metal or as a deployment.

damemi replied:
I mean, if I set a targetExe will this loop continue searching all pids even after the target pid is found (unnecessarily)? Would you maybe want to update L123 to something like:

```go
pids[pid] = exe
if strings.Contains(string(exeFullPath), target.ExePath) {
  break
}
```

damemi commented:
> @edeNFed This PR is working on baremetal and as sidecar for a pod. But when i run same as a deployment in k8s, am getting error below error. Can you please help with same.

@dineshg13 could you share the deployment yaml you're trying?

dineshg13 commented:
@damemi I have converted our test case from side car to a deployment. 

`auto-instrumentation - pod `
```
apiVersion: v1
kind: Pod
metadata:
  name: "auto-instrumentation"
  namespace: default
  labels:
    app: "auto-instrumentation"
spec:
  containers:
    - name: auto-instrumentation
      image: otel-go-instrumentation
      imagePullPolicy: IfNotPresent
      env:
        - name: OTEL_GO_AUTO_TARGET_EXE
          value: /sample-app/main
        - name: OTEL_EXPORTER_OTLP_ENDPOINT
          value: "http://test-opentelemetry-collector:4317"
        - name: OTEL_SERVICE_NAME
          value: "sample-app"
        - name: OTEL_PROPAGATORS
          value: "tracecontext,baggage"
      resources: {}
      securityContext:
        runAsUser: 0
        runAsGroup: 0
        capabilities:
          add:
            - SYS_PTRACE
        privileged: true
      volumeMounts:
        - mountPath: /sys/kernel/debug
          name: kernel-debug
        - mountPath: /proc
          name: hostproc
  volumes:
    - name: kernel-debug
      hostPath:
        path: /sys/kernel/debug
    - name: hostproc
      hostPath:
        path: /proc
  restartPolicy: Always
---
```
our app
```
apiVersion: v1
kind: Pod
metadata:
  name: "sample-app"
  namespace: default
  labels:
    app: "sample-app"
spec:
  containers:
    - name: sample-app
      image: sample-app
      imagePullPolicy: IfNotPresent
      command: ["/bin/sh", "-c"]
      args: ["/sample-app/main"]
      ports:
        - containerPort: 8080
          name: http
  restartPolicy: Always
---
```

edeNFed commented:
@dineshg13 please try to use `hostPID: true` instead of mounting the `/proc` directory

edeNFed commented:
Looks good @dineshg13 
I think we should merge #211 before this one, so we will avoid overwriting different BPF structures when instrumenting multiple processes.

damemi commented on cli/main.go:
```diff
@@ -35,56 +58,49 @@ func main() {
+			return
+		}
+	}
+	r, err := orchestrator.New(ctx, targetArgs, traceExporter)
```
what if `targetArgs` is nil here?

damemi commented:
> @damemi i prefer to add an e2e test in a follow-up PR. This PR is already big and backward compatible with existing functionality. The current test framework has only k8s sidecar, I want to refactor the tests to support all 3 paths viz sidecar, deployment & bare-metal. Let me know your thoughts.

I think it would be a lot better to have some testing for the new functionality before merging. You should be able to use the workflow matrix to deploy a different set of manifests for the sample app + instrumentation. I'm happy to help if you have any questions

dineshg13 commented:
@damemi i added bare-metal tests to the workflow. Currently the tests are failing on service-name. This might require some discussion, I have added this agenda to next SIG meeting.

damemi commented:
@dineshg13 thanks, the bare metal test is a good idea and probably wouldn't fit with the kind tests since the setup is so different.

If it helps, I put together a commit https://github.com/damemi/opentelemetry-go-instrumentation/pull/10/commits/9ef0f4d6e5dd24454e21b9390cde1e0834300fd8 based on your branch that shows what I mean for the deployment tests. Feel free to pick this commit to this PR or I can push it to your branch if you'd like (not sure I have permission). The tests pass in a test PR https://github.com/damemi/opentelemetry-go-instrumentation/pull/10

Right now, I think this would be good enough to show everything working in the new feature. The fixtures can be shared between deployment+sidecar modes because we expect the output to be identical. But if we add more configurability (like service name), we should look at restructuring the fixtures. This can be part of a broader refactor for the e2e framework.

dineshg13 commented:
@damemi Thanks. I was able to cherry-pick the commit and push the same. Ideally, for both deployment & bare-metal we should run one instance of collector for multiple apps. We can do that refactor later as well. We should make the service configurable via yaml instead of env vars. We can discuss this tomorrow.

MrAlias commented:
Specification reference for service name: https://github.com/open-telemetry/opentelemetry-specification/tree/bb1aa87ed221c16fc915a9fbfbfa262b989e1f79/specification/resource/semantic_conventions#service

dineshg13 commented:
@damemi @MrAlias @edeNFed  The PR is ready for review. I changed the code to read `OTEL_SERVICE_NAME` from the process envars. I have made it mandatory, and auto-instrumenation would only monitor processes with `OTEL_SERVICE_NAME` defined. It also removes the need to have ignore-processes list. Let me know your thoughts.

edeNFed commented on pkg/orchestrator/orchestrator.go:
```diff
@@ -0,0 +1,169 @@
+
+			return nil
+		case d := <-i.deadProcess:
+			log.Logger.Info("process died cleaning up", "pid", d)
```
Need to clean BPFFS for dead processes. This can be done by calling `bpffs.Cleanup` with the dead PID

edeNFed commented:
@dineshg13 is this ready for another review?

dineshg13 commented:
@edeNFed yes.

MrAlias commented on CHANGELOG.md:
```diff
@@ -8,6 +8,8 @@ OpenTelemetry Go Automatic Instrumentation adheres to [Semantic Versioning](http
 
 ## [Unreleased]
 
+- Add ability to monitor multiple processes. ([#197](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/197))
```
```suggestion
### Added

- Add ability to monitor multiple processes. ([#197](https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/197))
```

MrAlias commented on pkg/process/ptrace/ptrace_linux.go:
```diff
@@ -99,7 +99,17 @@ func NewTracedProgram(pid int, logger logr.Logger) (*TracedProgram, error) {
+						retryCount[tid],
+						"limit",
+						threadRetryLimit,
+						"error",
+						err,
```
If this error is not `nil` shouldn't this be logged as an error?

MrAlias commented on pkg/process/ptrace/ptrace_linux.go:
```diff
@@ -208,5 +217,13 @@ func (p *TracedProgram) Step() error {
 // Mmap runs mmap syscall.
 func (p *TracedProgram) Mmap(length uint64, fd uint64) (uint64, error) {
-	return p.Syscall(syscall.SYS_MMAP, 0, length, syscall.PROT_READ|syscall.PROT_WRITE|syscall.PROT_EXEC, syscall.MAP_ANON|syscall.MAP_PRIVATE|syscall.MAP_POPULATE, fd, 0)
+	return p.Syscall(
+		syscall.SYS_MMAP,
+		0,
+		length,
+		syscall.PROT_READ|syscall.PROT_WRITE|syscall.PROT_EXEC,
+		syscall.MAP_ANON|syscall.MAP_PRIVATE|syscall.MAP_POPULATE,
+		fd,
+		0,
+	)
```
This change does not look related to the PR:

```suggestion
	return p.Syscall(syscall.SYS_MMAP, 0, length, syscall.PROT_READ|syscall.PROT_WRITE|syscall.PROT_EXEC, syscall.MAP_ANON|syscall.MAP_PRIVATE|syscall.MAP_POPULATE, fd, 0)
```

dineshg13 commented on pkg/process/discover.go:
```diff
@@ -91,30 +77,79 @@ func (a *Analyzer) findProcessID(target *TargetArgs) (int, error) {
+}
+
+func getEnvVars(pid int) (map[string]string, error) {
+	bytes, err := ioutil.ReadFile(fmt.Sprintf("/proc/%d/environ", pid))
```
Our ebpf auto instrumentation only works on linux systems and it needs `/proc` to be mounted. For Linux systems we support, this should be good.

dineshg13 commented on pkg/process/discover.go:
```diff
@@ -91,30 +77,79 @@ func (a *Analyzer) findProcessID(target *TargetArgs) (int, error) {
-func (a *Analyzer) Close() {
-	a.done <- true
+func (a *Analyzer) isGo(pid int) bool {
+	// TODO
```
removed.

dineshg13 commented on pkg/instrumentors/manager.go:
```diff
@@ -84,9 +82,9 @@ func (m *Manager) GetRelevantFuncs() map[string]interface{} {
+// filterUnusedInstrumentors filterers Instrumentors whose functions are
 // already instrumented out of the Manager.
-func (m *Manager) FilterUnusedInstrumentors(target *process.TargetDetails) {
+func (m *Manager) filterUnusedInstrumentors(target *process.TargetDetails) {
```
Added

dineshg13 commented on pkg/instrumentors/runner.go:
```diff
@@ -91,12 +93,12 @@ func (m *Manager) load(target *process.TargetDetails) error {
 }
 
 func (m *Manager) cleanup() {
-	close(m.incomingEvents)
```
good catch. Added it back

dineshg13 commented on Makefile:
```diff
@@ -132,9 +135,10 @@ fixtures/%:
 	fi
-	helm install test -f .github/workflows/e2e/k8s/collector-helm-values.yml opentelemetry-helm-charts/charts/opentelemetry-collector
+	helm install test -f .github/workflows/e2e/k8s/$(FIXTURE_MODE)/collector-helm-values.yml opentelemetry-helm-charts/charts/opentelemetry-collector
+	sleep 5
```
Fixture tests were failing locally without sleep. The next waits for the container to be ready, but sometimes the container isn't created and fails.

dineshg13 replied:
We already have `kubectl` wait for condition below. Wait for condition is failing when i run this locally.

MrAlias replied:
The issue with adding sleep is that is going to be brittle (it may work on you machine, but fail on others) and they accumulate.

I'm surprised the `wait` command would fail, but I'm also curious if the `rollout` command works better for you.

dineshg13 replied:
Converted to `until`

dineshg13 commented on pkg/process/analyze.go:
```diff
@@ -144,9 +144,6 @@ func (a *Analyzer) Analyze(pid int, relevantFuncs map[string]interface{}) (*Targ
 		EndAddr:   addr + mapSize,
 	}
 
-	if err != nil {
-		return nil, err
-	}
```
we already have `err` check above. after the check `err` isn't assigned

dineshg13 commented on pkg/orchestrator/orchestrator.go:
```diff
@@ -0,0 +1,176 @@
+	ctx context.Context,
+	targetArgs *process.TargetArgs,
+	exporter sdktrace.SpanExporter,
+) (Interface, error) {
```
Updated. Added Service struct

dineshg13 commented on pkg/orchestrator/orchestrator.go:
```diff
@@ -0,0 +1,176 @@
+	ctx context.Context,
+	targetArgs *process.TargetArgs,
+	exporter sdktrace.SpanExporter,
+) (Interface, error) {
```
I am using https://github.com/segmentio/golines for formatting. is it not recommended ?

MrAlias replied:
It does not match the rest of the project or the other OTel Go projects. I would not use that formater if it produces this syntax.

MrAlias commented on CHANGELOG.md:
```diff
@@ -8,20 +8,27 @@ OpenTelemetry Go Automatic Instrumentation adheres to [Semantic Versioning](http
+- `pkg/process/` Analyzer struct `Close` method is removed.
 - The function signature of `"go.opentelemetry.io/auto/offsets-tracker/downloader".DownloadBinary` has changed.
-  It now has an additional flag indicating whether it'll build a dummy app for Go stdlib packages or not. ([#256]https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/256)
+  It now has an additional flag indicating whether it'll build a dummy app for Go stdlib packages or not. ([#256]<https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/256>)
```
```suggestion
  It now has an additional flag indicating whether it'll build a dummy app for Go stdlib packages or not. ([#256](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/256))
```

MrAlias commented on CHANGELOG.md:
```diff
@@ -8,20 +8,27 @@ OpenTelemetry Go Automatic Instrumentation adheres to [Semantic Versioning](http
 
 - Fix context propagation across different goroutines. ([#118](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/118))
-- The offset tracker can once again build binaries for the Go stdlib. ([#256]https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/256)
+- The offset tracker can once again build binaries for the Go stdlib. ([#256]<https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/256>)
```
```suggestion
- The offset tracker can once again build binaries for the Go stdlib. ([#256](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/256))
```

MrAlias commented on CHANGELOG.md:
```diff
@@ -8,20 +8,27 @@ OpenTelemetry Go Automatic Instrumentation adheres to [Semantic Versioning](http
+
 ### Changed
 
+- `pkg/instrumentors` Manager struct method  FilterUnusedInstrumentors is made private.
+- `pkg/opentelemtry/` NewController method now accepts ControllerSettings param.
+- `pkg/process/` Analyzer struct `Close` method is removed.
```
```suggestion
- The `FilterUnusedInstrumentors` method of  `"go.opentelemetry.io/auto/pkg/instrumentors".Manager` is removed. ([#208](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/208))
- The function signature of `"go.opentelemetry.io/auto/pkg/opentelemtry".NewController` has changed.
  It now accepts `context.Context` and `ControllerSettings` parameters. ([#208](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/208))
- The `Close` method of `"go.opentelemetry.io/auto/pkg/process".Analyzer` is removed. ([#208](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/208))
```

MrAlias commented on pkg/orchestrator/orchestrator.go:
```diff
@@ -0,0 +1,176 @@
+}
+
+// New creates a new Implementation of orchestrator Service.
+func New(
+	ctx context.Context,
+	targetArgs *process.TargetArgs,
+	exporter sdktrace.SpanExporter,
+) (*Service, error) {
```
```suggestion
func New(ctx context.Context, targetArgs *process.TargetArgs, exporter sdktrace.SpanExporter) (*Service, error) {
```

MrAlias commented on pkg/opentelemetry/controller.go:
```diff
@@ -98,35 +83,22 @@ func (c *Controller) convertTime(t int64) time.Time {
-	}
-
-	ctx := context.Background()
+func NewController(
+	ctx context.Context,
+	settings ControllerSetting,
+) (*Controller, error) {
```
```suggestion
func NewController(ctx context.Context, settings ControllerSetting) (*Controller, error) {
```

MrAlias commented:
Please rebase these changes into `internal/pkg` (changed in https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/282). The `pkg` directory is going to be removed in https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/302 given that package is no longer used.

MrAlias commented on cli/main.go:
```diff
@@ -34,30 +30,15 @@ func main() {
-	log.Logger.V(0).Info("starting instrumentors...")
-	if err = inst.Run(ctx); err != nil && err != errors.ErrInterrupted {
-		log.Logger.Error(err, "instrumentation crashed")
+	log.Logger.V(0).Info("starting Go OpenTelemetry Agent ...")
```
```suggestion
	log.Logger.V(0).Info("starting OpenTelemetry Go Agent ...")
```

MrAlias commented on instrumentation.go:
```diff
@@ -65,64 +62,62 @@ func NewInstrumentation(opts ...InstrumentationOption) (*Instrumentation, error)
-	}
 
-	mngr, err := instrumentors.NewManager(ctrl)
+	ctx := contextWithSigterm(context.Background())
```
The change in context handling seems incorrect to me. We want users to still be able to cancel a `Run` meaning they need to be able to pass a context to that method.

dineshg13 replied:
👍  Added context to `Run` method

MrAlias replied:
The context added to `Run` is still this context. It should not be. The creator of an `Instrumentor` needs to be the one creating and passing this context.

dineshg13 replied:
Added it as arg.

MrAlias commented on internal/pkg/orchestrator/orchestrator.go:
```diff
@@ -0,0 +1,221 @@
+
+// New creates a new Implementation of orchestrator Service.
+func New(
+	ctx context.Context,
```
This should be passed directly to `Run`.

MrAlias commented on internal/pkg/orchestrator/service.go:
```diff
@@ -0,0 +1,209 @@
+)
+
+// Service is responsible for managing all instrumentation.
+type Service struct {
```
Why is service defined across both this file and `orchestrator.go`? Can you merge the type definition and its methods into a single file?

MrAlias commented on internal/pkg/orchestrator/orchestrator.go:
```diff
@@ -0,0 +1,221 @@
+}
+
+// New creates a new Implementation of orchestrator Service.
+func New(
```
```suggestion
func NewService(
```

The created type is not named `Orchestrator`.

MrAlias commented on internal/pkg/orchestrator/orchestrator.go:
```diff
@@ -0,0 +1,221 @@
+		processch:   make(chan *pidServiceName, 10),
+		deadProcess: make(chan int, 10),
+		managers:    make(map[int]*instrumentors.Manager),
+		pidTicker:   time.NewTicker(2 * time.Second).C,
```
This ticker looks to never have its `Stop` method called and its resources are never appropriately released.

dineshg13 replied:
Calling `Stop`  when context is done.

MrAlias commented on internal/pkg/orchestrator/service.go:
```diff
@@ -0,0 +1,209 @@
+	})
+}
+
+func WithMonitorAll(monitorAll bool) ServiceOpt {
+	return fnOpt(func(c Service) Service {
+		c.monitorAll = monitorAll
+		return c
+	})
+}
+
+func WithExporter(expoter sdktrace.SpanExporter) ServiceOpt {
+	return fnOpt(func(c Service) Service {
+		c.exporter = expoter
+		return c
+	})
+}
+
+func WithPID(pid int) ServiceOpt {
+	return fnOpt(func(c Service) Service {
+		c.pid = pid
+		return c
+	})
+}
+
+func WithVersion(version string) ServiceOpt {
```
These need comments documenting them.

MrAlias commented on internal/pkg/orchestrator/service.go:
```diff
@@ -0,0 +1,209 @@
+	})
+}
+
+func WithMonitorAll(monitorAll bool) ServiceOpt {
```
This seems to always be `true`. Is there a reason we need it as an option?

dineshg13 replied:
removed it for now. We can add in future, if we would like to have a version of auto version without ability to monitor all services.

MrAlias commented on internal/pkg/orchestrator/service.go:
```diff
@@ -0,0 +1,209 @@
+	pidTicker   <-chan time.Time
+}
+
+type ServiceOpt interface {
```
Missing documentation.

MrAlias commented on internal/pkg/orchestrator/service.go:
```diff
@@ -0,0 +1,209 @@
+	return serviceName
+}
+
+func (s Service) Validate() error {
```
Missing documentation.

MrAlias commented on internal/pkg/orchestrator/service.go:
```diff
@@ -0,0 +1,209 @@
+}
+
+func (s Service) Validate() error {
+	if s.pid != 0 {
+		return validatePID(s.pid)
+	}
+
+	if s.monitorAll {
+		return nil
+	}
+	if s.serviceName == "" {
+		return fmt.Errorf("serviceName is nil")
+	}
+
+	if s.exePath == "" {
+		return fmt.Errorf("execPath is nil")
+	}
```
This will only evaluate at most one of these conditionals. Is this intentional?

dineshg13 replied:
Yes. It is the current behaviour as well.

MrAlias commented on instrumentation.go:
```diff
@@ -65,64 +62,62 @@ func NewInstrumentation(opts ...InstrumentationOption) (*Instrumentation, error)
 
-	mngr, err := instrumentors.NewManager(ctrl)
+	ctx := contextWithSigterm(context.Background())
+	log.Logger.V(0).Info("Establishing connection to OTLP receiver ...")
+	otlpTraceClient := otlptracegrpc.NewClient(
+		otlptracegrpc.WithDialOption(grpc.WithUserAgent(autoinstUserAgent)),
+	)
+	traceExporter, err := otlptrace.New(ctx, otlpTraceClient)
```
Should this be passed as an option instead?

It seems like this could continue to be setup in the controller. It could be changed in a targeted follow up PR instead.

I think changing the exporter setup in this PR is going to bloat this PR. Especially since this doesn't seem like the appropriate way to expose the option to the user.

dineshg13 replied:
We want to use same exporter for monitoring all the processes. If we do inside the controller, we will be creating on grpc connection for each process.

MrAlias commented on internal/pkg/instrumentors/runner.go:
```diff
@@ -28,27 +27,24 @@ import (
 
 // Run runs the event processing loop for all managed Instrumentors.
-func (m *Manager) Run(ctx context.Context, target *process.TargetDetails) error {
+func (m *Manager) Run(target *process.TargetDetails) error {
```
Removing the `ctx` arg means this cannot be canceled and shut down appropriately. I don't think that is an appropriate change, right?

RonFed commented on internal/pkg/orchestrator/service.go:
```diff
@@ -0,0 +1,209 @@
+//
+// If OTEL_GO_AUTO_TARGET_EXE is defined it will take precedence over any value
+// passed here.
+func WithTarget(path string) ServiceOpt {
```
Is there a way to avoid this duplication with the functions in `instrumentation.go` ?

dineshg13 replied:
I would like remove `instrumentation.go` . I will do so in future PRs. After orchestrator implementation, there is no need for `instrumentation.go` . I didn't want to introduce breaking changes in this PR.

dineshg13 commented on internal/pkg/orchestrator/orchestrator.go:
```diff
@@ -0,0 +1,221 @@
+	if s.pid != 0 {
+		s.processch <- &pidServiceName{
+			pid:         s.pid,
+			serviceName: s.serviceName,
```
Yes. That is intent.

MrAlias commented on instrumentation.go:
```diff
@@ -65,64 +63,62 @@ func NewInstrumentation(opts ...InstrumentationOption) (*Instrumentation, error)
 // Run starts the instrumentation.
-func (i *Instrumentation) Run(ctx context.Context) error {
-	return i.manager.Run(ctx, i.target)
+func (i *Instrumentation) Run() error {
+	return i.orchestrator.Run(i.ctx)
```
```suggestion
func (i *Instrumentation) Run(ctx context.Context) error {
	return i.orchestrator.Run(ctx)
```

dineshg13 commented:
There are too many breaking changes. It has become harder to update the PR. Closing the effort.
